File: scripts/dumpPrompts.ts
// Renders the shared parity fixtures through the REAL prompts.ts and writes the
// (system, user) pairs to finetune/work/parity/ts-dump.json. The Python side
// (finetune/scripts/dump_prompts.py) renders the same fixtures through the port;
// finetune/scripts/parity_check.py diffs the two. Run from maestro-open/:
//   npx vite-node scripts/dumpPrompts.ts
// NOTE: the runtime ` /no_think` system suffix (quirks.ts) is appended HERE so both
// dumps show exactly what the model sees.

import { readFileSync, writeFileSync, mkdirSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { join, dirname } from 'node:path';
import {
  classifyPrompt,
  expandPrompt,
  refinePrompt,
  coveragePrompt,
  teachPrompt,
  suggestionsPrompt,
  assessPrompt,
  syncPrompt,
  completionPrompt,
} from '../src/engine/milestone/prompts';
import type { Milestone } from '../src/engine/milestone/types';

const here = dirname(fileURLToPath(import.meta.url));
const repoRoot = join(here, '..', '..');
const fixtures = JSON.parse(readFileSync(join(repoRoot, 'finetune', 'data', 'parity-fixtures.json'), 'utf8'));

const NO_THINK = ' /no_think';
const pair = (p: { system: string; user: string }) => ({ system: p.system + NO_THINK, user: p.user });
// teach/assess/sync fixtures carry only the fields the prompt builders read.
const asMilestone = (m: { description: string; context: unknown[] }) => m as unknown as Milestone;

const out: Record<string, unknown[]> = {
  classify: fixtures.classify.map((f: { lessonTitle: string; goal: string; depth: number; maxDepth: number }) =>
    pair(classifyPrompt(f.lessonTitle, f.goal, f.depth, f.maxDepth))),
  expand: fixtures.expand.map((f: { goal: string }) => pair(expandPrompt(f.goal))),
  refine: fixtures.refine.map((f: { goals: string[]; draftSteps: string[] }) => pair(refinePrompt(f.goals, f.draftSteps))),
  coverage: fixtures.coverage.map((f: { id: string; statement: string }) => pair(coveragePrompt(f))),
  teach: fixtures.teach.map((f: {
    milestone: { description: string; context: unknown[] };
    justAdvanced: boolean;
    bridge?: { completedTitle: string; lastStudentMessage: string; mastered: boolean };
    attempts?: number;
    rails?: Record<string, unknown>;
  }) => pair(teachPrompt(asMilestone(f.milestone), f.justAdvanced, f.bridge, f.attempts ?? 0, f.rails))),
  suggestions: fixtures.suggestions.map((f: { tutorReply: string; milestoneTitle: string }) =>
    pair(suggestionsPrompt(f.tutorReply, f.milestoneTitle))),
  assess: fixtures.assess.map((f: { milestone: { description: string; context: unknown[] } }) =>
    pair(assessPrompt(asMilestone(f.milestone)))),
  sync: fixtures.sync.map((f: { completed: { description: string; context: unknown[] }; remaining: Array<{ id: string; description: string }> }) =>
    pair(syncPrompt(asMilestone(f.completed), f.remaining as unknown as Milestone[]))),
  completion: fixtures.completion.map((f: { title: string }) => pair(completionPrompt({ title: f.title } as never))),
};

const outDir = join(repoRoot, 'finetune', 'work', 'parity');
mkdirSync(outDir, { recursive: true });
writeFileSync(join(outDir, 'ts-dump.json'), JSON.stringify(out, null, 2));
console.log(`wrote ${join(outDir, 'ts-dump.json')}`);

File: finetune/scripts/dump_prompts.py
"""Render the shared parity fixtures through the Python port (gen/prompts.py) and
write finetune/work/parity/py-dump.json. Counterpart of maestro-open/scripts/dumpPrompts.ts.
Run from finetune/:  work/venv/bin/python scripts/dump_prompts.py
"""

import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(ROOT))

from gen import prompts as P  # noqa: E402

fixtures = json.loads((ROOT / "data" / "parity-fixtures.json").read_text())


def pair(system, user):
    return {"system": P.sys(system), "user": user}


out = {
    "classify": [
        pair(*P.classify_prompt(f["lessonTitle"], f["goal"], f["depth"], f["maxDepth"]))
        for f in fixtures["classify"]
    ],
    "expand": [pair(*P.expand_prompt(f["goal"])) for f in fixtures["expand"]],
    "refine": [pair(*P.refine_prompt(f["goals"], f["draftSteps"])) for f in fixtures["refine"]],
    "coverage": [pair(*P.coverage_prompt(f["statement"])) for f in fixtures["coverage"]],
    "teach": [
        pair(
            *P.teach_prompt(
                f["milestone"],
                f["justAdvanced"],
                bridge=f.get("bridge"),
                attempts=f.get("attempts", 0),
                rails=f.get("rails"),
            )
        )
        for f in fixtures["teach"]
    ],
    "suggestions": [
        pair(*P.suggestions_prompt(f["tutorReply"], f["milestoneTitle"]))
        for f in fixtures["suggestions"]
    ],
    "assess": [pair(*P.assess_prompt(f["milestone"])) for f in fixtures["assess"]],
    "sync": [pair(*P.sync_prompt(f["completed"], f["remaining"])) for f in fixtures["sync"]],
    "completion": [pair(*P.completion_prompt(f["title"])) for f in fixtures["completion"]],
}

out_dir = ROOT / "work" / "parity"
out_dir.mkdir(parents=True, exist_ok=True)
(out_dir / "py-dump.json").write_text(json.dumps(out, indent=2, ensure_ascii=False))
print(f"wrote {out_dir / 'py-dump.json'}")
